### PR TITLE
Allow empty

### LINKED ISSUE #149: Way to specify that a list or collection is required, ie have at least one valid entry
With the required method being deprecated it is more difficult to specify a schema with a required list or collection, that has to have at least one valid entry. Currently this has to be done through a validator, which is kind of technical for Jo the engineer.  Propose to add built inn option to specify this in the schema definition.

### PR DESCRIPTION
**Pull request by markusdregi**
Resolves #149 

~~Note: We should probably merge #152 first...~~

### CODE REVIEW

**Review by jondequinor — COMMENTED** [review]
Maybe check that I got the rebase right?

Otherwise, this looks OK

File: configsuite/schema.py [hunk 1]
```diff
@@ -76,9 +76,17 @@ def _check_required_not_default(schema_level):
     return True
 
 
+@configsuite.validator_msg("Only variable lenght containers can specify allowempty")
+def _check_allowempty_only_variable_containers(schema_level):
+    level_type = schema_level.get(MK.Type)
+    const_len = isinstance(level_type, types.BasicType) or level_type == types.NamedDict
```
Comment (jondequinor) [inline configsuite/schema.py:82]: What does `const_len` mean? And why vary between `isinstance` and `==` for comparator? A comment would maybe clarify that.

File: configsuite/validator.py [hunk 2]
```diff
@@ -75,6 +75,9 @@ def _validate(self, config, schema):
         if self._apply_validators and valid:
             valid &= self._context_validation(config, schema)
 
+        if self._apply_validators and valid:
+            valid &= self._length_validation(config, schema)
```
Comment (jondequinor) [inline configsuite/validator.py:79]: It's too late now, but should we not stick to logical operators for boolean values?

`valid = valid and self._length_validation(config, schema)`

Comment (markusdregi) [reply configsuite/validator.py:79]: I'll stick with the above in this PR due to my earlier doings right above. However, I'll definitively accept any patch removing the bitwise operators. Or, if you expand your argument of why we should do so, I might do it myself. But either way, I think it belongs in a separate PR 👍

File: configsuite/schema.py (hunk 1 above)
Comment (markusdregi) [reply configsuite/schema.py:82]: As the name of the function hints towards, we now distinguish between variable length containers (list and dict) and constant length data types (basic type and named dict). The idea is that only the former is allowed to specify `allowempty`, as the notion has no meaning for the later.

Comment (markusdregi) [reply configsuite/schema.py:82]: I agree that the mixture between `isinstance` and `==` is unfortunate. It is due to the fact that `BasicType` is a meta type in configsuite (and is implemented as a class), while `NamedDict` is a concrete type (implemented as an instance of `Container`). This is an implementation detail that is leaking too far and we could very well abstract the difference in implementation away in a meaningful manner!

Comment (markusdregi) [reply configsuite/schema.py:82]: My hope was that the validation message, the function name and the variable name (which should perhaps have been `is_const_length`?) would be sufficient documentation. But I'm more than willing to add a description as well :) If you mind my answers above clarifying; do you have any suggestions for a helpful comment?

**Comment by markusdregi** [comment]
This PR should also contain an updated version of the release notes and documentation. I'll try to add that ASAP

**Comment by markusdregi** [comment]
@jondequinor I've rebased, added documentation and updated the dev release notes. I think it is good to go 🚀

**Review by jondequinor — COMMENTED** [review]
Should using `AllowEmpty` silence the `DeprecationWarning: Specifying whether elements are required directly is deprecated. Use ConfigSuite(..., deduce_required=True)` warning?

We're deprecating `deduce_required` and suggesting `AllowEmpty` plus layering instead.

File: configsuite/schema.py [hunk 3]
```diff
@@ -76,9 +76,17 @@ def _check_required_not_default(schema_level):
     return True
 
 
+@configsuite.validator_msg("Only variable lenght containers can specify allowempty")
```
Comment (jondequinor) [inline configsuite/schema.py:79]:
```suggestion
@configsuite.validator_msg("Only variable length containers can specify AllowEmpty")
```

File: configsuite/schema.py (hunk 1 above)
Comment (jondequinor) [reply configsuite/schema.py:82]: No, this becomes clearer after reading the docs/loading the context of the PR.

File: docs/source/user_guide/getting_started.rst [hunk 4]
```diff
@@ -568,6 +568,37 @@ in order for a configuration with ``None`` to pass.
...
 
 
+Allow Empty
+-----------
+
+Occasionally one would like to require variable length containers (lists and
+dicts) to have at least one element. This can be implemented with a validator
+on the container (see :ref:`Validators-section`). However, this feature have
+been requested multiple times and as of such we've decided to implement
+explicit support for it without having to implement your own validator. Note
```
Comment (jondequinor) [inline docs/source/user_guide/getting_started.rst:578]:
```suggestion
on the container (see :ref:`Validators-section`). However, as this feature have
been requested multiple times, we've decided to implement
explicit support for it without having to implement your own validator. Note
```

**Comment by markusdregi** [comment]
> Should using `AllowEmpty` silence the `DeprecationWarning: Specifying whether elements are required directly is deprecated. Use ConfigSuite(..., deduce_required=True)` warning?
> 
> We're deprecating `deduce_required` and suggesting `AllowEmpty` plus layering instead.

I'm suspecting that you are confusing `AllowNone` with `AllowEmpty`, as it is `AllowNone` that will aid in removing the `Required` option. The intended process from a consumer perspective regarding the notion of required is outlined here in the [docs](https://github.com/equinor/configsuite/blob/master/docs/source/user_guide/deprecated_behaviour.rst). I just wanted to get his one in first and then resolve that matter afterwards.

tldr; This is a stand-alone feature that has nothing to do with the deprecation of requiring content.

File: tests/test_basic_types.py [hunk 5]
```diff
@@ -309,3 +185,13 @@ def test_invalid_boolean_result_type(self):
...
             with self.assertRaises(TypeError):
                 configsuite.types.BooleanResult(elem, "incorrect", "type")
+
+    def test_allow_empty_basic_type_invalid(self):
+        for val in [True, False]:
+            schema = {MK.Type: configsuite.types.String, MK.AllowEmpty: val}
+            with self.assertRaises(ValueError) as err:
+                configsuite.ConfigSuite("fdsfds", schema)
+            self.assertIn(
+                "Only variable lenght containers can specify allowempty is false",
```
Comment (markusdregi) [inline tests/test_basic_types.py:195]:
```suggestion
                "Only variable length containers can specify AllowEmpty is false",
```

File: tests/test_containers.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+            schema = {
+                MK.Type: configsuite.types.NamedDict,
+                MK.AllowEmpty: val,
+                MK.Content: {"a": {MK.Type: configsuite.types.Integer}},
+            }
+
+            with self.assertRaises(ValueError) as err:
+                configsuite.ConfigSuite("fdsfds", schema)
+            self.assertIn(
+                "Only variable lenght containers can specify allowempty is false",
```
Comment (markusdregi) [inline tests/test_containers.py:209]:
```suggestion
                "Only variable length containers can specify AllowEmpty is false",
```

**Review by jondequinor — APPROVED** [review]
LGTM! But please squash the commits before merging.
